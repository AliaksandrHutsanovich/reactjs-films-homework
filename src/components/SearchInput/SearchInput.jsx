import React, { memo, useState } from 'react';
import PropTypes from 'prop-types';
import { Input } from 'antd';
import clsx from 'clsx';

import '!style-loader!css-loader!sass-loader!antd/dist/antd.css'; //eslint-disable-line
import getStateSetter from '../../utils/getStateSetter';
import styles from './SearchInput.scss';

const { Search } = Input;

const SearchInput = ({
  size,
  onSearch,
}) => {
  const [inputState, setState] = useState('blured');

  return (
    <Search
      className={clsx(styles.input, { [styles.focused]: inputState === 'focused' })}
      allowClear
      size={size}
      onSearch={onSearch}
      onFocus={getStateSetter(setState, 'focused')}
      onBlur={getStateSetter(setState, 'blured')}
      enterButton
    />
  );
};

SearchInput.propTypes = {
  size: PropTypes.oneOf(['large', 'middle', 'small']),
  onSearch: PropTypes.func.isRequired,
};

SearchInput.defaultProps = {
  size: 'middle',
};

export default memo(SearchInput);
