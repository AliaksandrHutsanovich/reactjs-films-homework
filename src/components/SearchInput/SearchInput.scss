.input {
  width: 300px;
  background-color: transparent;
  border-radius: 3px;

  > input {
    background-color: transparent;
  }

  > span > span:last-child {
    background-color: transparent;

    > button {
      background-color: transparent;
      border: 1px solid white;
      width: 32px;
      border-radius: 0 3px 3px 0;
      align-items: center;
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }
}

.focused {
  background-color: rgb(93, 155, 212);

  > span > span:last-child > button {
    background-color: rgb(93, 155, 212);
    border: 1px solid rgb(93, 155, 212);
  }
}
