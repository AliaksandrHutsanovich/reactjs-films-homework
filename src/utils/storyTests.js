const storyTests = (stories, renderer) => {
  Object.keys(stories).forEach((storyKey) => {
    if (typeof stories[storyKey] !== 'function') {
      return;
    }
    it(`${storyKey} should be rendered properly`, () => {
      renderer.render(stories[storyKey]());
      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });
};

export default storyTests;
