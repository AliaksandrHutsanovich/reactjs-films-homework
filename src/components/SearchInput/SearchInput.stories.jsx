import React from 'react';
import { action } from '@storybook/addon-actions';
import SearchInput from './SearchInput';
import image from '../../assets/images/story-background.png';

export default {
  component: SearchInput,
  title: 'components/SearchInput',
  decorators: [
    (story) => (
      <div
        style={{
          background: `url(${image}) no-repeat`,
          padding: '20px',
          width: '800px',
          height: '800px',
        }}
      >
        {story()}
      </div>
    ),
  ],
};

export const search = () => (
  <SearchInput onSearch={action('search has started')} />
);
