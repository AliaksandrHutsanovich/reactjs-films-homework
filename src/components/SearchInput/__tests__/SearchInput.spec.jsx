import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import storyTests from '../../../utils/storyTests';
import * as stories from '../SearchInput.stories';
import SearchInput from '../SearchInput';

// every function should be prefixed mock
const mockFun = jest.fn();

jest.mock('../../../utils/getStateSetter', () => (fun, val) => () => {
  mockFun(val);
});

describe('<SearchInput />', () => {
  const renderer = new ShallowRenderer();

  storyTests(stories, renderer);

  it('onFocus and onBlur should be called', () => {
    renderer.render(<SearchInput onSearch={() => {}} />);
    const result = renderer.getRenderOutput();

    result.props.onFocus();
    expect(mockFun).toHaveBeenCalledWith('focused');

    result.props.onBlur();
    expect(mockFun).toHaveBeenCalledWith('blured');
  });
});
