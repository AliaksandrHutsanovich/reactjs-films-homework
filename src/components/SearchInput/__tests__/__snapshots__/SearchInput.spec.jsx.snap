// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> search should be rendered properly 1`] = `
<ForwardRef
  allowClear={true}
  className="input"
  enterButton={true}
  onBlur={[Function]}
  onFocus={[Function]}
  onSearch={[Function]}
/>
`;
